buildSrc/src/main/groovy/org/elasticsearch/gradle/VersionCollection.groovy

```diff
@@ -25,159 +25,293 @@ import java.util.regex.Matcher
+ *   -- caveat 2: the largest released minor branch before the unreleased minor should also be a snapshot
+ *   -- caveat 3: if the current version is a different major than the previous rules apply to major - 1 of the current version
+ *
+ * Please note that the caveat's also correspond with the 4 types of snapshots.
+ * - Caveat 0 - always maintenanceBugfixSnapshot.
+ * - Caveat 1 - This is tricky. If caveat 3 applies, the highest matching value is nextMinorSnapshot, if there is another it is the stagedMinorSnapshot.
+ *              If caveat 3 does not apply then the only possible value is the stagedMinorSnapshot.
+ * - Caveat 2 - always nextBugfixSnapshot
+ * - Caveat 3 - this only changes the applicability of Caveat 1
+ *
+ * Notes on terminology:
+ * - The case for major+1 being released is accomplished through the isReleasableBranch value. If this is false, then the branch is no longer
+ *   releasable, meaning not to test against any snapshots.
+ * - Released is defined as having > 1 suffix-free version in a major.minor series. For instance, only 6.2.0 means unreleased, but a
+ *   6.2.0 and 6.2.1 mean that 6.2.0 was released already.
  */
 class VersionCollection {
 
     private final List<Version> versions
+    Version nextMinorSnapshot
+    Version stagedMinorSnapshot
+    Version nextBugfixSnapshot
+    Version maintenanceBugfixSnapshot
+    final Version currentVersion
+    private final TreeSet<Version> versionSet = new TreeSet<>()
+
+    // When we roll 8.0 its very likely these will need to be extracted from this class
     private final boolean buildSnapshot = System.getProperty("build.snapshot", "true") == "true"
+    private final boolean isReleasableBranch = true
 
     /**
-     * Construct a VersionCollection from the lines of the Version.java file.
+     * Construct a VersionCollection from the lines of the Version.java file. The basic logic for the following is pretty straight forward.
+
      * @param versionLines The lines of the Version.java file.
      */
     VersionCollection(List<String> versionLines) {
 
         List<Version> versions = []
+        // This class should be converted wholesale to use the treeset
 
         for (final String line : versionLines) {
             final Matcher match = line =~ /\W+public static final Version V_(\d+)_(\d+)_(\d+)(_alpha\d+|_beta\d+|_rc\d+)? .*/
             if (match.matches()) {
                 final Version foundVersion = new Version(
                         Integer.parseInt(match.group(1)), Integer.parseInt(match.group(2)),
-                        Integer.parseInt(match.group(3)), (match.group(4) ?: '').replace('_', '-'), false, null)
-
-                if (versions.size() > 0 && foundVersion.onOrBeforeIncludingSuffix(versions[-1])) {
-                    throw new GradleException("Versions.java contains out of order version constants:" +
```

Comment:
I can't see this check any more, but ISTR it's important. Maybe it's elsewhere, I can't remember right now.

Reply:
Its a treemap now, so it sorts it. there is not really a way i can check. i think it was important so we didnt have to sort it in the code below where that original check was.